File: RB101/lesson_2/rock_paper_scissors/rock_paper_scissors.rb
VALID_CHOICES = {
  'rock' => 'r',
  'paper' => 'p',
  'scissors' => 's',
  'spock' => 'sp',
  'lizard' => 'l'
}

WINNING_MOVES = {
  'rock' => %w(scissors lizard),
  'lizard' => %w(paper spock),
  'spock' => %w(scissors rock),
  'scissors' => %w(paper lizard),
  'paper' => %w(rock spock)
}

def prompt(message)
  Kernel.puts("=> #{message}")
end

def clear_screen
  system('clear') || system('cls')
end

def reset_score
  reset_score = {
    player: 0,
    computer: 0
  }
end

def increment_score(score, key)
  score[key] += 1
end

def win?(first, second)
  WINNING_MOVES[first].include?(second)
end

def result(player, computer)
  if win?(player, computer)
    'player'
  elsif win?(computer, player)
    'computer'
  else
    'tie'
  end
end

def set_score(winner, score)
  if winner == 'player'
    increment_score(score, :player)
  elsif winner == 'computer'
    increment_score(score, :computer)
  end
end

def print_score(winner, score)
  both_score_message = <<~MSG
    You've won #{score[:player]} games,
    computer has won #{score[:computer]} games
  MSG

  if winner == 'player'
    prompt("You won! #{both_score_message}")
  elsif winner == 'computer'
    prompt("Computer won! #{both_score_message}")
  else
    prompt("Tie!")
  end
end

def retrieve_choice
  loop do
    choose = <<~MSG
      Choose one: #{VALID_CHOICES.keys.join(', ')}
      You can also enter the shortened versions: #{VALID_CHOICES.values.join(', ')}
    MSG

    prompt(choose)

    choice = gets.chomp

    if VALID_CHOICES.include?(choice)
      return choice
    elsif VALID_CHOICES.value?(choice)
      return choice = VALID_CHOICES.key(choice)
    else
      prompt("Invalid choice - please try again")
    end
  end
end

def random_choice
  VALID_CHOICES.to_a.sample[0]
end

def display_choices(user_choice, computer_choice)
  prompt("You chose #{user_choice}; computer chose: #{computer_choice}")
end

def display_grand_winner(score)
  prompt("#{score.key(GOAL_WINS)} has won #{GOAL_WINS} games - #{score.key(GOAL_WINS)} wins!")
end

def someone_grand_winner(score)
  score.value?(GOAL_WINS)
end

def play_again?
  loop do
    valid_answers = ['y', 'yes', 'no', 'n']
    prompt("Do you want to play again? Y/N")
    answer = gets.chomp.downcase
    next prompt("Invalid entry, try again") unless valid_answers.include?(answer)
    break false if answer == 'n' || answer == 'no'
    break true if answer == 'y' || answer == 'yes'
  end
end

def display_goodbye
  prompt("Thank you for playing. Good bye!")
end

score = {
  player: 0,
  computer: 0
}

GOAL_WINS = 5

loop do
  prompt("Welcome to rock-paper-scissors!")
  prompt("First to five game wins wins the round!")

  user_choice = retrieve_choice

  computer_choice = random_choice

  display_choices(user_choice, computer_choice)

  winner = result(user_choice, computer_choice)

  set_score(winner, score)

  print_score(winner, score)

  break display_grand_winner(score) if someone_grand_winner(score)

  break unless play_again?
end

display_goodbye

File: js101/lesson_2/rock_paper_scissors.js
const readline = require("readline-sync");
const VALID_CHOICES = {
  r: "rock",
  p: "paper",
  sc: "scissors",
  l: "lizard",
  sp: "spock"
};

const WINNING_COMBOS = {
  rock:     ["scissors",    "lizard"],
  paper:    ["rock",        "spock"],
  scissors: ["paper",       "lizard"],
  lizard:   ["spock",       "paper"],
  spock:    ["scissors",    "rock"]
};

const GOAL_WINS = 3;

let scores = {
  player: 0,
  computer: 0
};

function playerWins(choice, computerChoice) {
  return WINNING_COMBOS[choice].includes(computerChoice);
}

function displayChoices(choice, computerChoice) {
  prompt(`You chose ${choice}, computer chose ${computerChoice}`);
}

function result(choice, computerChoice) {
  if (playerWins(choice, computerChoice)) {
    return "player";
  } else if (choice === computerChoice) {
    return "tie";
  } else {
    return "computer";
  }
}

function prompt(message) {
  console.log(`=> ${message}`);
}

function incrementScore(scores, winner) {
  scores[winner] += 1;
}

function setScore(scores, winner) {
  incrementScore(scores, winner);
}

function printScore (winner) {
  if (winner === "player") {
    prompt("You win!");
  } else if (winner === "computer") {
    prompt("You lose");
  } else {
    prompt("It's a tie!");
  }
  prompt(`You've won ${scores.player} games so far.`);
  prompt(`computer has won ${scores.computer} games so far`);
}

function randomChoice() {
  let validChoicesVals = Object.values(VALID_CHOICES);
  let randomIndex = Math.round(Math.random() * validChoicesVals.length);
  if (randomIndex === validChoicesVals.length) {
    randomIndex = validChoicesVals.length - 1;
  }
  return validChoicesVals[randomIndex];
}

function retrieveChoice() {
  let validChoicesKeys = Object.keys(VALID_CHOICES);
  let validChoicesVals = Object.values(VALID_CHOICES);
  prompt(`Choose one: ${validChoicesVals.join(", ")}`);
  let choice = readline.question();

  while (!validChoicesKeys.includes(choice) &&
   !validChoicesVals.includes(choice)) {
    if (choice === "s") {
      prompt("two options begin with s - please enter the first and second letter");
    } else {
      prompt("That's not a valid choice");
    }
    choice = readline.question();
    if (choice.length <= 2) choice = VALID_CHOICES[choice];
  }
  if (choice.length <= 2) choice = VALID_CHOICES[choice];
  return choice;
}

function someoneGrandWinner() {
  return (scores.player === GOAL_WINS) || (scores.computer === GOAL_WINS);
}

function displayGrandWinner() {
  let winner;
  if (scores.player === GOAL_WINS) {
    winner = "player";
  } else {
    winner = "computer";
  }
  prompt(`best of three reached - ${winner} wins!`);
  prompt("Game over!");
}

function retrievePlayAgainChoice() {
  prompt("Do you want to play again (y/n)?");
  let answer = readline.question().toLowerCase();
  while (answer[0] !== "n" && answer[0] !== "y") {
    prompt("Please enter 'y' or 'n'.");
    answer = readline.question().toLowerCase();
  }
  return answer[0];
}

while (true) {
  let choice = retrieveChoice();

  let computerChoice = randomChoice();

  displayChoices(choice, computerChoice);

  let winner = result(choice, computerChoice);

  setScore(scores, winner);

  printScore(winner);

  if (someoneGrandWinner()) {
    displayGrandWinner();
  }

  let playAgainChoice = retrievePlayAgainChoice();

  if (playAgainChoice !== "y") break;
}